base_name_search_improved/models/ir_model.py

```diff
@@ -0,0 +1,261 @@
+        .search([("model", "=", str(self._name))])
+        .name_search_domain
+    )
+    if name_search_domain:
+        return literal_eval(name_search_domain)
+    return []
+
+
+def _extend_name_results(self, domain, results, limit):
+    result_count = len(results)
+    if result_count < limit:
+        domain += [("id", "not in", results)]
+        rec_ids = self._search(
+            domain,
+            limit=limit - result_count,
+        )
+        results.extend(rec_ids)
+    return results
+
+
+def patch_name_search():
+    @api.model
+    def _name_search(
+        self, name="", domain=None, operator="ilike", limit=100, order=None
+    ):
+        # Perform standard name search
+        res = _name_search.origin(
+            self,
+            name=name,
+            domain=domain,
+            limit=limit,
+            order=order,
+        )
+        if name and _get_use_smart_name_search(self.sudo()) and operator in ALLOWED_OPS:
+            # _name_search.origin is a query, we need to convert it to a list
+            res = self.browse(res).ids
+            limit = limit or 0
+
+            # we add domain
+            args = domain or [] + _get_name_search_domain(self.sudo())
+
+            # Support a list of fields to search on
+            all_names = _get_rec_names(self.sudo())
+            base_domain = args or []
+            # Try regular search on each additional search field
+            for rec_name in all_names[1:]:
+                domain = [(rec_name, operator, name)]
+                res = _extend_name_results(self, base_domain + domain, res, limit)
+            # Try ordered word search on each of the search fields
+            for rec_name in all_names:
+                domain = [(rec_name, operator, name.replace(" ", "%"))]
+                res = _extend_name_results(self, base_domain + domain, res, limit)
+            # Try unordered word search on each of the search fields
+            # we only perform this search if we have at least one
+            # separator character
+            # also, if have raise the limit we skeep this iteration
+            if " " in name and len(res) < limit:
+                domain = []
+                for word in name.split():
+                    word_domain = []
+                    for rec_name in all_names:
+                        word_domain = (
+                            word_domain and ["|"] + word_domain or word_domain
+                        ) + [(rec_name, operator, word)]
+                    domain = (domain and ["&"] + domain or domain) + word_domain
+                res = _extend_name_results(self, base_domain + domain, res, limit)
+
+        return res
+
+    return _name_search
+
+
+class Base(models.AbstractModel):
+    _inherit = "base"
+
+    # TODO perhaps better to create only the field when enabled on the model
+    smart_search = fields.Char(
+        compute="_compute_smart_search", search="_search_smart_search", translate=False
+    )
```

Comment:
@rov-adhoc how did you try to implement @StefanRijnhart's suggestion? I'd expect it to work to override `_setup_fields`, check via sql if the model has the functionality enabled, and if so, add the field to `self._fields`.

Just setting translate to False can't work because it's not about the translations of *data* from the field, but about things like its label.